src/uct/ib/base/ib_iface.c

```diff
@@ -667,6 +667,60 @@ uct_ib_iface_roce_is_reachable(const uct_ib_device_gid_info_t *local_gid_info,
     return ret;
 }
 
+static int
+uct_ib_iface_is_same_device(const uct_ib_address_pack_params_t *params,
+                            uct_ib_iface_t *iface)
+{
+    const union ibv_gid *gid = &iface->gid_info.gid;
+    int res = 1;
+
+    if ((params->flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) !=
+        uct_ib_iface_is_roce(iface)) {
+        return 0;
+    }
+
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) {
+        res &= !memcmp(gid->raw, params->gid.raw, sizeof(params->gid.raw));
+    } else {
+        res &= (uct_ib_iface_port_attr(iface)->lid == params->lid);
+        if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID &&
+            params->gid.global.interface_id != 0) {
```

Comment:
why need to check != 0 ?

Reply:
in case we did not pack it in address:
https://github.com/openucx/ucx/blob/f02f1578a3225cdca97148e57048842373fcc9e1/src/uct/ib/base/ib_iface.c#L497

Reply:
isn't it checked by `params->flags & UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID` ?

Reply:
seems that this flag is appended always, regardless to the value of `UCT_IB_ADDRESS_FLAG_IF_ID`:

https://github.com/openucx/ucx/blob/f02f1578a3225cdca97148e57048842373fcc9e1/src/uct/ib/base/ib_iface.c#L491

Reply:
so let's try to fix the unpack